### PR TITLE
Use hash table for counter allocated size

### PR DESCRIPTION
**Pull request by michalbiesek**
- reduce the overhead of access same counter variable by multiple threads simultaneously
- calculate hash based on thread_id
- provide atomic store implementation
- export atomic API to static functions

- [ ] Find the optimum value for buckets - make it configurable?

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/621)

### CODE REVIEW

**Review by bratpiorka — COMMENTED** [review]
LGTM, do you have any performance numbers that shows how the overhead is reduced?

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/621#-:-MXqPBoY-KnMzQh3T5xT:ba0f5pg)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @michalbiesek)

File: src/memkind_memtier.c [hunk 1]
```diff
@@ -64,11 +74,38 @@ struct memtier_memory {
...
+    uint64_t x = (uint64_t)pthread_self();
+    x += 0x9e3779b97f4a7c15;
+    x = (x ^ (x >> 30)) * 0xbf58476d1ce4e5b9;
+    x = (x ^ (x >> 27)) * 0x94d049bb133111eb;
+    return (x ^ (x >> 31)) & (THREAD_BUCKETS - 1);
+}
+
+static inline void increment_alloc_size(unsigned kind_id, size_t size)
+{
+    unsigned b_id = t_hash_64();
```
Comment (jschmieg) [inline src/memkind_memtier.c:101]: bucket_id instead of b_id would be a much more meaningful name

**Review by PatKamin — COMMENTED** [review]
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/621)* status: all files reviewed, 1 unresolved discussion (waiting on @michalbiesek)

File: src/memkind_memtier.c (hunk 1 above)
Comment (michalbiesek) [reply src/memkind_memtier.c]: Done.

**Review by bratpiorka — COMMENTED** [review]
:lgtm:

> Reviewed 1 of 1 files at r2, 1 of 1 files at r3, 2 of 2 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, 1 of 1 files at r9.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/621#-:-MZC1Upq2gkjB8A7bLvU:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @jschmieg)

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 4 of 4 files at r10, 2 of 2 files at r11, 1 of 1 files at r12, 1 of 1 files at r13, 1 of 1 files at r14, 1 of 1 files at r15, 1 of 1 files at r16.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/621)* status: all files reviewed, 1 unresolved discussion (waiting on @jschmieg)

**Review by PatKamin — COMMENTED** [review]
:lgtm:

> Reviewed 4 of 4 files at r10.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/621#-:-MZDIsPGBF1RKpIYRJzJ:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @jschmieg)

**Pull request merged**